File: 프로그래머스/Py/LV2. 오픈채팅방.py
# 해쉬+문자열

def solution(record):

    user = {}
    messages = []
    res = []

    for r in record:
        tmp = r.split()
        if tmp[0] == "Enter":
            user[tmp[1]] = tmp[2]
            messages.append([tmp[1], "Enter"])
        elif tmp[0] == "Leave":
            messages.append([tmp[1], "Leave"])
        elif tmp[0] == "Change":
            user[tmp[1]] = tmp[2]

    for m in messages:
        id, oper = m[0], m[1]
        if oper == "Enter":
            res.append(user[id]+"님이 들어왔습니다.")
        elif oper == "Leave":
            res.append(user[id]+"님이 나갔습니다.")

    return res

File: 프로그래머스/JS/LV2. 오픈채팅방.js
function solution(record) {
  let ans = [];

  let set = {};
  for (let re of record) {
    let [order, id, name] = re.split(" ");
    if (name) {
      set[id] = name;
    }
  }

  for (let re of record) {
    let [order, id, name] = re.split(" ");
    if (order === "Enter") {
      ans.push(`${set[id]}님이 들어왔습니다.`);
    } else if (order === "Leave") {
      ans.push(`${set[id]}님이 나갔습니다.`);
    }
  }

  return ans;
}
